packages/react-menu/src/components/Menu/useMenu.tsx

```diff
@@ -78,19 +78,28 @@ export const useMenu = (props: MenuProps, ref: React.Ref<HTMLElement>, defaultPr
   const [open, setOpen] = useControllableValue(state.open, state.defaultOpen);
   // TODO fix useControllableValue typing
   state.open = open !== undefined ? open : state.open;
+  const { onOpenChange } = state;
   state.setOpen = React.useCallback(
-    (...args) => {
-      setOpen(...args);
+    (e, shouldOpen) => {
+      setOpen(prevOpen => {
+        // More than one event (mouse, focus, keyboard) can request the popup to close
+        // We assume the first event is the correct one
+        if (prevOpen !== shouldOpen) {
+          onOpenChange?.(e, { open: shouldOpen });
+        }
+
+        return shouldOpen;
+      });
     },
-    [setOpen],
+    [setOpen, onOpenChange],
   );
 
   useMenuSelectableState(state);
   useMenuPopup(state);
   useOnClickOutside({
     element: document,
     refs: [state.menuPopupRef, triggerRef],
-    callback: () => setOpen(false),
+    callback: e => state.setOpen((e as unknown) as React.MouseEvent<HTMLElement>, false),
```

Comment:
Correct me, if I am wrong, but this one is a DOM element. I don't think that it's correct to use typecasts there. Let's rather define proper types for it as consumers don't know anything about our internals:

```tsx
setOpen: (e: MouseEvent | React.KeyboardEvent<HTMLElement> | React.MouseEvent<HTMLElement>, open: boolean) => void;
```

Reply:
that's fair, I was lazy

Reply:
```suggestion
    callback: e => state.setOpen(e, false),
```

?